### PR TITLE
stylo: Add glue for contain property

### PR DESCRIPTION
**Pull request by canova**
In gecko, `size` and `content` values aren't implemented yet. So I had to disable these parts with mako. We will need to update the property and the glue once these are implemented in gecko.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix [Bug 1354998](https://bugzilla.mozilla.org/show_bug.cgi?id=1354998)

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16467)

### CODE REVIEW

**Comment by canaltinova** [comment]
https://treeherder.mozilla.org/#/jobs?repo=try&revision=986c3818087206cfdcb73a1c3ae342aa35e98406

File: components/style/properties/longhand/box.mako.rs [hunk 1]
```diff
@@ -2122,9 +2131,11 @@ ${helpers.single_keyword("transform-style",
             if self.contains(STRICT) {
                 return dest.write_str("strict")
             }
-            if self.contains(CONTENT) {
-                return dest.write_str("content")
```
Comment (heycam) [inline components/style/properties/longhand/box.mako.rs]: Hmm, looking at this, I realize that the existing serialization (and storage for specified/computed values) is wrong.  The spec says that the computed value is the same as the specified value, which means that `contains: layout style paint` must serialize as `layout style paint` and not as `content`.

The current behavior will probably cause some Gecko test failures, so I think you'll need to fix that.  Maybe we can just make `STRICT` and `CONTENT` be separate bits, but when we parse those values, we also set the specific bits they correspond to.  (That's what Gecko does.)  Then `to_css` should work properly without any further changes.

File: components/style/properties/properties.mako.rs [hunk 2]
```diff
@@ -2203,6 +2203,7 @@ pub fn apply_declarations<'a, F, I>(device: &Device,
                     | LonghandId::AnimationName
                     | LonghandId::TransitionProperty
                     | LonghandId::XLang
+                    | LonghandId::Contain
```
Comment (heycam) [inline components/style/properties/properties.mako.rs]: Why is this needed?

**Review by heycam — DISMISSED** [review]

File: components/style/properties/gecko.mako.rs [hunk 3]
```diff
@@ -2235,6 +2235,36 @@ fn static_assert() {
...
+        use gecko_bindings::structs::NS_STYLE_CONTAIN_NONE;
+        use gecko_bindings::structs::NS_STYLE_CONTAIN_STRICT;
+        use gecko_bindings::structs::NS_STYLE_CONTAIN_LAYOUT;
+        use gecko_bindings::structs::NS_STYLE_CONTAIN_STYLE;
+        use gecko_bindings::structs::NS_STYLE_CONTAIN_PAINT;
+        use gecko_bindings::structs::NS_STYLE_CONTAIN_ALL_BITS;
+        use properties::longhands::contain;
+
+        self.gecko.mContain = if v.contains(contain::STRICT) {
+            NS_STYLE_CONTAIN_STRICT | NS_STYLE_CONTAIN_ALL_BITS
```
Comment (emilio) [inline components/style/properties/gecko.mako.rs]: On top of the bit @heycam spotted, I see that gecko splits `STRICT` and `ALL_BITS`, presumably for serialization purposes (so `contain: layout style paint` doesn't serialize as `contain: strict`). Can you verify it?

If so, we should ensure the serialization behavior is the same, and add an `ALL_PROPERTIES` bit to Servo too. Once that's done, we can test that the constants are the expected ones with unit tests (see the `assert_restyle_hints_match` test, and this can just become:

```rust
self.gecko.mContain = v.bits;
```

Does that sounds reasonable?

**Review by emilio — CHANGES_REQUESTED** [review]

File: components/style/properties/longhand/box.mako.rs (hunk 1 above)
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: Ouch, I was late, d'oh

File: components/style/properties/gecko.mako.rs (hunk 3 above)
Comment (emilio) [inline components/style/properties/gecko.mako.rs]: I only read the first @heycam comment, so please disregard this :)

**Comment by canaltinova** [comment]
Oh, I misunderstood the grammar in the glue. Thanks for catching that! Fixed the parsing/serialization problem and the glue.

@emilio I didn't wanted to change the bits to match gecko ones since they are not completely implemented in gecko yet. The gecko side change of this property can easily break the stylo build. (I can change if you still want :) )

File: components/style/properties/properties.mako.rs (hunk 2 above)
Comment (canaltinova) [inline components/style/properties/properties.mako.rs]: The bug in bugzilla says contain is an early property. So I added it here.

**Comment by emilio** [comment]
Instead of making it an early property, we should add a fixup to the `cascade` function (just like `our_writing_mode` and `parent_writing_mode` may affect `display` as well).

**Comment by emilio** [comment]
(That being said, the only display-related bit I see is https://drafts.csswg.org/css-contain/#becoming-formatting-context, and doesn't seem like it should change the computed value of `display`?)

**Comment by canaltinova** [comment]
Removed this from early properties and added fixups for overflow and display properties. Additional tests seems passing:
https://treeherder.mozilla.org/#/jobs?repo=try&revision=96ceb5ac0353645f3d52552804951fa28dc1d8a8

**Comment by canaltinova** [comment]
Hm, there are also 3 unexpected fails:
```
REFTEST TEST-UNEXPECTED-FAIL | file:///home/worker/workspace/build/tests/reftest/tests/layout/reftests/w3c-css/submitted/contain/contain-paint-clip-003.html | assertion count 2 is more than expected 0 assertions
REFTEST TEST-UNEXPECTED-FAIL | file:///home/worker/workspace/build/tests/reftest/tests/layout/reftests/w3c-css/submitted/contain/contain-paint-clip-004.html | assertion count 2 is more than expected 0 assertions
REFTEST TEST-UNEXPECTED-FAIL | file:///home/worker/workspace/build/tests/reftest/tests/layout/reftests/w3c-css/submitted/contain/contain-paint-clip-005.html | assertion count 2 is more than expected 0 assertions 
```

**Comment by emilio** [comment]
Those are actually passes (there are less assertions).

File: components/style/properties/longhand/box.mako.rs [hunk 4]
```diff
@@ -2093,14 +2096,21 @@ ${helpers.single_keyword("transform-style",
...
+            % endif
             const LAYOUT = 0x02,
             const STYLE = 0x04,
             const PAINT = 0x08,
-            const STRICT = 0x10,
-            const STRICT_BITS = SIZE.bits | LAYOUT.bits | STYLE.bits | PAINT.bits,
-            const CONTENT = 0x20,
-            const CONTENT_BITS = LAYOUT.bits | STYLE.bits | PAINT.bits,
+            % if product == "gecko":
+                const STRICT = 0x10,
```
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: The strict definition is the same for both, so let's move this out.

Also, this feature is gecko-only, so I'd just remove the product checks and remove `contains: content`.

**Review by emilio — CHANGES_REQUESTED** [review]

File: components/style/properties/longhand/box.mako.rs (hunk 4 above)
Comment (canaltinova) [inline components/style/properties/longhand/box.mako.rs]: I thought the same too but we'll eventually have to implement this property for servo and eventually gecko will implement `contains: content`. I think just removing the if's are better than implementing it again (well, it's trivial to add again, but still :) )

Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: I don't really think it's worth the dead code tbh.

**Comment by canaltinova** [comment]
Fair enough, removed the non-gecko parts.

File: components/style/properties/longhand/box.mako.rs [hunk 5]
```diff
@@ -2093,14 +2096,11 @@ ${helpers.single_keyword("transform-style",
...
-            const PAINT = 0x08,
-            const STRICT = 0x10,
-            const STRICT_BITS = SIZE.bits | LAYOUT.bits | STYLE.bits | PAINT.bits,
-            const CONTENT = 0x20,
-            const CONTENT_BITS = LAYOUT.bits | STYLE.bits | PAINT.bits,
+            const LAYOUT = 0x01,
+            const STYLE = 0x02,
+            const PAINT = 0x04,
+            const STRICT = 0x8,
+            const STRICT_BITS = LAYOUT.bits | STYLE.bits | PAINT.bits,
```
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: We still need the extra flag to indicate that `content` was specified, don't we?

Comment (canaltinova) [inline components/style/properties/longhand/box.mako.rs]: `content` is not implemented in gecko yet. `strict` is currently equivalent of `layout style paint`. We need to update these once they are implemented.

Comment (canaltinova) [inline components/style/properties/longhand/box.mako.rs]: I removed that flag when you said it's not worth the dead code

Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: NVM, my fault completely, I meant `STRICT`, I missed that one.

File: components/style/properties/longhand/box.mako.rs [hunk 6]
```diff
@@ -353,6 +353,7 @@ ${helpers.single_keyword("overflow-clip-box", "padding-box content-box",
 
 // FIXME(pcwalton, #2742): Implement scrolling for `scroll` and `auto`.
 ${helpers.single_keyword("overflow-x", "visible hidden scroll auto",
+                         extra_gecko_values="clip",
```
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: Is this meant to be part of this PR? If so, split it into another commit, otherwise, remove?

**Review by emilio — APPROVED** [review]
Looks good to me with that, thanks :)

File: components/style/properties/longhand/box.mako.rs (hunk 6 above)
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: Aand, it was already split. oh well, can't review so late in  the night.

Comment (canaltinova) [inline components/style/properties/longhand/box.mako.rs]: 😄  Btw, I needed to update overflow-{x,y} to `clip` if content is `paint` in `apply_declarations` function.
